src/core/domain.py

```
@@ -212,21 +213,40 @@ def spark_configurations(self) -> dict[str, str]:
         return items
 
 
-class ServiceAccount(StateBase):
+class ServiceAccount:
     """Class representing the service account managed by the Spark Integration Hub charm."""
 
-    def __init__(self, relation: Relation, component: Application):
-        super().__init__(relation, component)
+    def __init__(self, relation_data: SparkServiceAccountProviderData, relation_id: int):
+        self.relation_data = relation_data
+        self.relation_id = relation_id
 
     @property
     def service_account(self) -> str | None:
         """Return service account name."""
-        return self.relation_data.get("service-account", None)
+        return self.relation_data.fetch_my_relation_field(
+            relation_id=self.relation_id, field="service-account"
+        )
 
     @property
-    def namespace(self) -> str:
-        """Return the used namespace."""
-        return self.relation_data["namespace"]
+    def spark_properties(self) -> dict[str, str] | None:
+        """Return the set of Spark properties."""
+        field_data = self.relation_data.fetch_my_relation_field(
+            relation_id=self.relation_id, field="spark-properties"
+        )
+        if field_data is None:
+            return None
+
+        props = dict(json.loads(field_data))
```

**suggestion**: We can remove the `dict` call here, I don't think it is doing anything

---

My editor complained on `props.items()` citing props could be a list as well and thus may not have `items()` method. Wrapping it with `dict()` suppresses that error.

---

Ah yes, you are absolutely right, I forgot that `json.loads` could return a list